source/adios2/operator/compress/CompressPNG.cpp

```diff
@@ -0,0 +1,220 @@
+/*
+ * Distributed under the OSI-approved Apache License, Version 2.0.  See
+ * accompanying file Copyright.txt for details.
+ *
+ * CompressPNG.cpp
+ *
+ *  Created on: Jun 10, 2019
+ *      Author: William F Godoy godoywf@ornl.gov
+ */
+#include "CompressPNG.h"
+
+#include <cstring> // std::memset
+
+extern "C" {
+#include <png.h>
+}
+
+#include "adios2/helper/adiosFunctions.h"
+
+namespace adios2
+{
+namespace core
+{
+namespace compress
+{
+
+const std::map<std::string, uint32_t> CompressPNG::m_ColorTypes = {
+    {"PNG_COLOR_TYPE_GRAY", PNG_COLOR_TYPE_GRAY},
+    {"PNG_COLOR_TYPE_PALETTE", PNG_COLOR_TYPE_PALETTE},
+    {"PNG_COLOR_TYPE_RGB", PNG_COLOR_TYPE_RGB},
+    {"PNG_COLOR_TYPE_RGB_ALPHA", PNG_COLOR_TYPE_RGB_ALPHA},
+    {"PNG_COLOR_TYPE_GRAY_ALPHA", PNG_COLOR_TYPE_GRAY_ALPHA},
+    {"PNG_COLOR_TYPE_RGBA", PNG_COLOR_TYPE_RGBA},
+    {"PNG_COLOR_TYPE_GA", PNG_COLOR_TYPE_GA}};
+
+const std::map<std::string, std::set<uint32_t>> CompressPNG::m_BitDepths = {
+    {"PNG_COLOR_TYPE_GRAY", {1, 2, 4, 8, 16}},
+    {"PNG_COLOR_TYPE_PALETTE", {1, 2, 4, 8}},
+    {"PNG_COLOR_TYPE_RGB", {8, 16}},
+    {"PNG_COLOR_TYPE_RGB_ALPHA", {8, 16}},
+    {"PNG_COLOR_TYPE_GRAY_ALPHA", {8, 16}},
+    {"PNG_COLOR_TYPE_RGBA", {8, 16}},
+    {"PNG_COLOR_TYPE_GA", {8, 16}}};
+
+// PUBLIC
+CompressPNG::CompressPNG(const Params &parameters, const bool debugMode)
+: Operator("png", parameters, debugMode)
+{
+}
+
+size_t CompressPNG::Compress(const void *dataIn, const Dims &dimensions,
+                             const size_t elementSize,
+                             const std::string /*type*/, void *bufferOut,
+                             const Params &parameters, Params &info) const
+{
+    auto lf_Write = [](png_structp png_ptr, png_bytep data, png_size_t length) {
+        DestInfo *pDestInfo =
+            reinterpret_cast<DestInfo *>(png_get_io_ptr(png_ptr));
+        std::memcpy(pDestInfo->BufferOut + pDestInfo->Offset, data, length);
+        pDestInfo->Offset += length;
+    };
+
+    const std::size_t ndims = dimensions.size();
+    if (m_DebugMode)
+    {
+        if (ndims != 3 && ndims != 2)
+        {
+            throw std::invalid_argument(
+                "ERROR: image number of dimensions " + std::to_string(ndims) +
+                " is invalid, must be 2 {height,width*bytes_per_pixel} or 3"
+                " {height,width,bytes_per_pixel]} , in call to ADIOS2 PNG "
+                " compression\n");
```

Comment:
I'm worried about two things here: First, I feel like I'd want to see this error message independently of whether I was in debug mode. Second, the comment states that the meaning of the second argument depends on the presence of the third. I'd be concerned that people will write bugs using this.